analyzers/src/SonarAnalyzer.CSharp/Helpers/CSharpEquivalenceChecker.cs

```diff
@@ -22,16 +22,74 @@
 using System.Linq;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Extensions;
+using StyleCop.Analyzers.Lightup;
 
 namespace SonarAnalyzer.Helpers
 {
     internal static class CSharpEquivalenceChecker
     {
         public static bool AreEquivalent(SyntaxNode node1, SyntaxNode node2) =>
-            Common.EquivalenceChecker.AreEquivalent(node1, node2, (n1, n2) => SyntaxFactory.AreEquivalent(n1, n2));
+            Common.EquivalenceChecker.AreEquivalent(node1, node2, NodeComparator);
 
         public static bool AreEquivalent(SyntaxList<SyntaxNode> nodeList1, SyntaxList<SyntaxNode> nodeList2) =>
-            Common.EquivalenceChecker.AreEquivalent(nodeList1, nodeList2, (n1, n2) => SyntaxFactory.AreEquivalent(n1, n2));
+            Common.EquivalenceChecker.AreEquivalent(nodeList1, nodeList2, NodeComparator);
+
+        private static bool NodeComparator(SyntaxNode node1, SyntaxNode node2) =>
+            NullCheckState(node1) is { } nullCheck1
+            && NullCheckState(node2) is { } nullCheck2
+            && nullCheck1.Positive == nullCheck2.Positive
+                ? SyntaxFactory.AreEquivalent(nullCheck1.Expression, nullCheck2.Expression)
+                : SyntaxFactory.AreEquivalent(node1, node2);
+
+        private static NullCheck NullCheckState(SyntaxNode node, bool positive = true)
+        {
+            if (node is PrefixUnaryExpressionSyntax unary && unary.IsKind(SyntaxKind.LogicalNotExpression))
+            {
+                return NullCheckState(unary.Operand.RemoveParentheses(), !positive);
+            }
+            else if (node is BinaryExpressionSyntax binary && binary.IsAnyKind(SyntaxKind.EqualsExpression, SyntaxKind.NotEqualsExpression))
+            {
+                if (binary.IsKind(SyntaxKind.NotEqualsExpression))
+                {
+                    positive = !positive;
```

Comment:
`positive = !positive` is confusing logic because:
- it's a bit of a mind-twister because `!positive` is `negative` for the human mind - `isPositive` may be a better name
- the method has `bool positive = true` by default, and when seeing this I expect that "if I do not explicitly pass the parameter, it means it's positive", however it's not the case  because this is actually evaluated based on the syntax kind
 
And being a recursive function it makes it a bit harder to understand. 

Please add some documentation for the `NullCheckState` method functionality, specifying that in the first run the sign is evaluated inside. (it took me a while to understand it, which means it needs documentation to help the reader)